sql/moz-fx-data-shared-prod/mozilla_org_derived/ga_clients_v1/query.sql

```diff
@@ -0,0 +1,79 @@
+WITH historical_clients AS (
+  SELECT
+    *
+  FROM
+    mozilla_org_derived.ga_clients_v1
+),
+new_sessions AS (
+  SELECT
+    ga_client_id,
+    MIN(session_date) AS first_seen_date,
+    MAX(session_date) AS last_seen_date,
+    LOGICAL_OR(had_download_event) AS had_download_event,
+    STRUCT(
+      /* Geos */
+      MIN_BY(country, session_number) AS country,
+      MIN_BY(region, session_number) AS region,
+      MIN_BY(city, session_number) AS city,
+      /* Attribution */
+      MIN_BY(campaign_id, session_number) AS campaign_id,
+      MIN_BY(campaign, session_number) AS campaign,
+      MIN_BY(source, session_number) AS source,
+      MIN_BY(medium, session_number) AS medium,
+      MIN_BY(term, session_number) AS term,
+      MIN_BY(content, session_number) AS content,
+      /* Device */
+      MIN_BY(device_category, session_number) AS device_category,
+      MIN_BY(mobile_device_model, session_number) AS mobile_device_model,
+      MIN_BY(mobile_device_string, session_number) AS mobile_device_string,
+      MIN_BY(os, session_number) AS os,
+      MIN_BY(os_version, session_number) AS os_version,
+      MIN_BY(LANGUAGE, session_number) AS language,
+      MIN_BY(browser, session_number) AS browser,
+      MIN_BY(browser_version, session_number) AS browser_version,
+      /* Session */
+      MIN_BY(ga_session_id, session_number) AS ga_session_id
+    ) AS first_reported,
+  FROM
+    mozilla_org_derived.ga_sessions_v1
+  WHERE
+    -- Re-process three days, to account for late-arriving data
+    session_date
+    BETWEEN DATE_SUB(@session_date, INTERVAL 3 DAY)
+    AND @session_date
+  GROUP BY
+    ga_client_id
+  HAVING
+    ga_client_id IS NOT NULL
+)
+SELECT
+  ga_client_id,
+  -- Least and greatest return NULL if any input is NULL, so we coalesce each value first
+  LEAST(
+    COALESCE(_previous.first_seen_date, _current.first_seen_date),
+    COALESCE(_current.first_seen_date, _previous.first_seen_date)
+  ) AS first_seen_date,
+  GREATEST(
+    COALESCE(_previous.last_seen_date, _current.last_seen_date),
+    COALESCE(_current.last_seen_date, _previous.last_seen_date)
+  ) AS last_seen_date,
+  -- OR treats NULL differently for True/False. If any input is True, it will return True; otherwise it returns NULL.
+  COALESCE(
+    _previous.had_download_event
+    OR _current.had_download_event,
+    FALSE
+  ) AS had_download_event,
+  -- We take the current data for first_reported only if we don't already have first_reported data, or if we're backfilling
+  -- If equal, prefer new data
```

Comment:
This is the case where:
1. We ran `ga_sessions_v1` for day N
2. We run `ga_clients_v1` for day N
3. The next day (N+1), we backfilled the session data from `ga_sessions_v1` for day N
4. We run `ga_clients_v1` for day N+1 (which also includes that backfilled data from day N)

The second time it ran, it may have found some session information in day N+1 for a session that started in day N. That session would have already been present in `ga_sessions_v1`, but it was incomplete. Now it will be complete. We want to use the more complete session information to determine the `first_reported` information for that client.